### PR TITLE
Make the NEWOBJ API narrower

### PR DESCRIPTION
**Pull request by eightbitraptor**
Since the introduction of variable width allocation with `RVARGC` there are a lot of different `*NEWOBJ*` macros. Currently there are:

* `RB_RVARGC_NEWOBJ_OF`
* `RB_RVARGC_EC_NEWOBJ_OF`
* `RB_NEWOBJ_OF`, an alias of `RB_RVARGC_NEWOBJ_OF`
* `RB_EC_NEWOBJ_OF`, an alias of `RB_RVARGC_EC_NEWOBJ_OF`
* `NEWOBJ_OF`, an alias of `RB_RVARGC_NEWOBJ_OF`
* `RVARGC_NEWOBJ_OF`, an alias of `RB_RVARGC_NEWOBJ_OF`

This PR merges `RB_RVARGC_NEWOBJ_OF`,`RB_RVARGC_EC_NEWOBJ_OF`, `RVARGC_NEWOBJ_OF`, `RB_NEWOBJ_OF` and `NEWOBJ_OF` into a single macro that takes an execution context as an argument (which can be `NULL`; if so then the current execution context is found and used using `GET_EC()`). 

The resulting single macro has been named `NEWOBJ_OF` to reflect that it should now be the only way of creating new objects.

Both `RB_NEWOBJ_OF` and `NEWOBJ_OF` have seperate implementations that are part of the public API exposed in `include/ruby/internal/newobj.h` so they are available to extension authors. These have not been modified in any way.

Note that implementing this PR required `gc.h` to include `vm_core.h`. This was not possibly because `vm_core.h` included `shape.h`, which included `gc.h` creating a circular dependency.

To address this I have removed the `shape_list`, `next_shape_id` and `root_shape` members from `rb_vm_struct` and instead promoted them to their own type: `rb_shape_tree_t`. This is defined as a global object accessible using `rb_shape_tree_ptr` and the `GET_SHAPE_TREE()` function macro. 

In practice, this means that `GET_VM()->shape_list` etc are now replaced with `GET_SHAPE_TREE()->shape_list`.

### CODE REVIEW

**Comment by nobu** [comment]
In the commit "Remove RB_RVARGC_EC_NEWOBJ_OF and RB_EC_NEWOBJ_OF", "teh" seems a typo.

File: internal/gc.h [hunk 1]
```diff
@@ -141,25 +142,14 @@ struct rb_objspace; /* in vm_core.h */
...
+                   rb_wb_protected_newobj_of((ec) ? (ec) : GET_EC(), (c), (f) & ~FL_WB_PROTECTED, s) : \
                    rb_wb_unprotected_newobj_of((c), (f), s))
 
-/* optimized version of NEWOBJ() */
-#define RB_NEWOBJ_OF(var, T, c, f) RB_RVARGC_NEWOBJ_OF(var, T, c, f, RVALUE_SIZE)
-
-#define RB_EC_NEWOBJ_OF(ec, var, T, c, f) RB_RVARGC_EC_NEWOBJ_OF(ec, var, T, c, f, RVALUE_SIZE)
-
+#define RB_NEWOBJ_OF(var, T, c, f) RVARGC_NEWOBJ_OF(var, T, c, f, RVALUE_SIZE, NULL)
 #define NEWOBJ_OF(var, T, c, f) RB_NEWOBJ_OF((var), T, (c), (f))
```
Comment (peterzhu2118) [inline internal/gc.h:151]: Perhaps out of scope of this PR, but it might be nice to get rid of these two macros since we can essentially use `RVARGC_NEWOBJ_OF` (and pass in the size of the type we're trying to allocate). This might be beneficial for some allocators since not all objects are of `RVALUE_SIZE`.

File: mjit_c.rb [hunk 2]
```diff
@@ -439,32 +439,6 @@ def C.rb_cref_t
   def C.rb_execution_context_struct
     @rb_execution_context_struct ||= CType::Struct.new(
       "rb_execution_context_struct", Primitive.cexpr!("SIZEOF(struct rb_execution_context_struct)"),
-      vm_stack: [CType::Pointer.new { self.VALUE }, Primitive.cexpr!("OFFSETOF((*((struct rb_execution_context_struct *)NULL)), vm_stack)")],
```
Comment (peterzhu2118) [inline mjit_c.rb:442]: Why was this file changed?

**Review by peterzhu2118 — COMMENTED** [review]
I'm not a huge fan of passing in the EC to `RVARGC_NEWOBJ_OF` since the EC is just an implementation detail of the current GC, so in the future we should think about how we can get rid of passing the EC into the GC.

File: include/ruby/internal/core/rbasic.h [hunk 3]
```diff
@@ -46,6 +46,8 @@
     RBIMPL_CAST((int)(sizeof(VALUE[RBIMPL_RVALUE_EMBED_LEN_MAX]) / (sizeof(T))))
 /** @endcond */
 
+#define RVALUE_SIZE (sizeof(struct RBasic) + sizeof(VALUE[RBIMPL_RVALUE_EMBED_LEN_MAX]))
```
[the remark this answers is not in the document]
Comment (eightbitraptor) [reply include/ruby/internal/core/rbasic.h:49]: This is needed by both `internal/gc.h` and `vm_core.h`. Now that `vm_core.h` doesn't implicitly include `internal/gc.h` I had to make this available somewhere to both.

I hadn't initially appreciated that this change has made it public visible thought, thanks for pointing that out. I'll make it private.

File: ractor_core.h [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#include "internal/gc.h"
```
[the remark this answers is not in the document]
Comment (eightbitraptor) [reply ractor_core.h:1]: because `ractor_core.h` is relying on some functions defined inside `internal/gc.h` which were previously included implicitly through the inclusion of `vm_core.h`. Now that `vm_core.h` doesn't include `internal/gc.h` anymore we need to specify this dependancy explicitly.

File: mjit_c.rb (hunk 2 above)
Comment (eightbitraptor) [reply mjit_c.rb:442]: I applied the diff from https://github.com/Shopify/ruby/actions/runs/4313674348/jobs/7525759736 in order to get the build passing again.

I don't know much about mjit, or the bindgen code - but the `rb_execution_context_struct` has changed (to extract the shape tree), but it hasn't changed this much.

File: ext/socket/extconf.rb [hunk 5]
```diff
@@ -699,5 +699,7 @@ def %(s) s || self end
       "not needed"
     end
   end
+
+  $VPATH << '$(topdir)' << '$(top_srcdir)'
```
[the remark this answers is not in the document]
Comment (eightbitraptor) [reply ext/socket/extconf.rb:703]: The socket extensions `rubysocket.h` pulls in the "private" `include/gc.h`, which now depends on `vm_core.h`. `vm_core.h` pulls in `id.h`

when `tool/update-deps` generates the dependencies for the makefiles, it generates the line for `id.h` to be based on `VPATH`, which is configured in the `extconf.rb` for each of the extensions. By default `VPATH` does not include the actual source directory of the current Ruby so the dependency fails to resolve and linking fails.

We need to append the `topdir` and `top_srcdir` to VPATH to have the dependancy picked up correctly (and I believe we need both of these to cope with in-tree and out-of-tree builds).

I copied this from the approach taken in https://github.com/ruby/ruby/blob/master/ext/objspace/extconf.rb#L3

File: internal/gc.h (hunk 1 above)
Comment (eightbitraptor) [reply internal/gc.h:151]: Agreed. I didn't attempt this for this PR because these macros are also defined in the public `newobj.h` header. I'll likely attempt to remove/deprecate these in a future PR.

File: mjit_c.rb (hunk 2 above)
Comment (eightbitraptor) [reply mjit_c.rb]: This turned out to be a bug in the `mjit`/`rjit` bindgen code - when we're generating the bindings we grab all the nodes of the same type - we can have multiple nodes of the same type if there are multiple declarations in different header files, as there are with `rb_execution_context_struct` - but the code that generates the bindings uses the last node instance to find the struct members.

It looks like my code has changed the order in which the nodes are discovered so that the last entry is now just an empty declaration, instead of the actual declaration, and so rjit bindgen thinks it has no members.

Fixed by @k0kubun in https://github.com/ruby/ruby/commit/309dd50a01203ac107c12e5a962e6aa6add75529 so that now we search the node list for the declaration that contains the member definitions.

File: ext/socket/extconf.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (eightbitraptor) [reply ext/socket/extconf.rb:703]: For reference: I've added this explanation to the commit message too, in case this comes up for other extensions in the future.

File: internal/gc.h (hunk 1 above)
Comment (peterzhu2118) [reply internal/gc.h:151]: Isn't that a reason to remove this internally first? This means that there's a name conflict with `newobj.h` which will result in behavior that differs depending on the load order.

File: iseq.h [hunk 6]
```diff
@@ -325,6 +325,9 @@ VALUE rb_iseq_defined_string(enum defined_type type);
 /* vm.c */
 VALUE rb_iseq_local_variables(const rb_iseq_t *iseq);
 
+#include "shape.h"
```
Comment (peterzhu2118) [inline iseq.h:328]: Can you move this include to the top of the file?

File: internal/gc.h [hunk 7]
```diff
@@ -141,25 +142,14 @@ struct rb_objspace; /* in vm_core.h */
...
 
-#define RVALUE_SIZE (sizeof(struct RBasic) + sizeof(VALUE[RBIMPL_RVALUE_EMBED_LEN_MAX]))
-
-#define RB_RVARGC_NEWOBJ_OF(var, T, c, f, s) \
-  T *(var) = (T *)(((f) & FL_WB_PROTECTED) ? \
-                   rb_wb_protected_newobj_of((c), (f) & ~FL_WB_PROTECTED, s) : \
-                   rb_wb_unprotected_newobj_of((c), (f), s))
-
-#define RB_RVARGC_EC_NEWOBJ_OF(ec, var, T, c, f, s) \
+#define RVARGC_NEWOBJ_OF(var, T, c, f, s, ec) \
```
[the remark this answers is not in the document]
Comment (eightbitraptor) [reply internal/gc.h:145]: Yeah, I completely agree - My thinking was that I'd try and reduce the scope of the API first and then deal with the extra parameters in a separate PR, although I have no massive preference either way. 

We're already passing in the `ec` anyway when we use `RB_RVARGC_EC_NEWOBJ_OF` so the worst thing this PR is doing is making an existing problem more obvious.

File: internal/gc.h [hunk 8]
```diff
@@ -141,25 +142,11 @@ struct rb_objspace; /* in vm_core.h */
...
```
Comment (peterzhu2118) [inline internal/gc.h:145]: Nit: if we're getting rid of `NEWOBJ_OF`, can we rename `RVARGC_NEWOBJ_OF` to just `NEWOBJ_OF`?

**Comment by ko1** [comment]
> I'm not a huge fan of passing in the EC to `RVARGC_NEWOBJ_OF` since the EC is just an implementation detail of the current GC, so in the future we should think about how we can get rid of passing the EC into the GC.

The background to pass the `ec` is here:

* `ec` is needed for Ractor systems (from 3.0)
* Getting `ec` from TLS (thread-local storage) can have overhead
* Therefore I made a plan to migrate all C-API to have the first parameter `ec` (`rbc_str_new(c, ...)` for example, in this case `c` is "context"). It is like `mrb_` APIs of mruby.
* This is why `Primitive` for builtin methods receive the `ec` parameter from VM.

However, now C-compiler specific TLS feature (`__thread` on gcc for example https://gcc.gnu.org/onlinedocs/gcc/Thread-Local.html) provides nice performance on current Ruby (as I observed).

So if you measured at least microbenchmarks the following:

* with passing `ec` (current macro)
* without passing `ec` (and `GET_EC()` everytime)

and if we don't have a difference, I agree to avoid passing the `ec`.

**Comment by ko1** [comment]
> (which can be NULL; if so then the current execution context is found and used using GET_EC()).

For me, this kind of branch should be avoided to reduce branches.

**Comment by eightbitraptor** [comment]
I got around to looking at these benchmarks recently, and to summarise the situation so far:

I refactored the NEWOBJ api's to reduce complexity at the call sites. This has resulted in an api that requires us to either pass an `ec` or `NULL` to the `NEWOBJ_OF` internal macro.

@peterzhu2118 had a concern that requiring call sites to pass the `ec` was leaking GC abstractions.

@ko1 was concerned that _not_ passing the `ec` and instead relying on `GET_EC` might be a performance problem because of the overhead incurred by accessing a thread local every time.

Additionally @ko1 was concerned that the extra branch point required now in the `NEWOBJ_OF` macro might also cause a performance impact.

I built a micro-benchmark, designed to produce many `putstring` instructions ran it across several variations of this feature.

This is the benchmark:

```
eval("def string_literal; foo = [#{ 10_000.times.map { "'test'" }.join(",") }]; nil; end")

before = Process.clock_gettime(Process::CLOCK_MONOTONIC)
10000.times { string_literal }
p Process.clock_gettime(Process::CLOCK_MONOTONIC) - before
```

These are the branches I used, and how they compare:

* `master` - passes EC explicitly in the cases of the `putstring` and `newarray` instructions. So that we don't have to use `GET_EC` to retreive the thread local ec variable.

* `mvh-narrow-newobj-api` - Based on `master` but with the newobj paths in `gc.c` and the `NEWOBJ_OF` macros refactored to reduce paths through the code. `ec` still must be explicitly passed, although this is now decided using a branch generated by the macro, rather than by using completely seperate code paths as in `master`.

* `mvh-narrow-newobj-api-macro-branch` - Same as `mvh-narrow-newobj-api`. but additionally removes the branch point, by using some pre-processor string concatenation and dynamic macro name generation to handle the branch at the pre-processor level.

* `mvh-narrow-newobj-api-remove-ec` - This builds on top of the refactors carried out in `mvh-narrow-newobj-api` but also completely removes the code path that passes the `ec` in the `putstring` and `newarray` instructions, relying instead on always getting the `ec` from the thread. This eliminates all the extra code paths required to check the presence of the `ec`.

## results

Screenshot 2023-03-28 at 17 21 29

|     | master     | mvh-narrow-newobj-api | mvh-narrow-newobj-api-remove-ec | mvh-narrow-newbj-api-macro-branch |
| --- | ---------- | --------------------- | ------------------------------- | --------------------------------- |
| 1   | 3.50582845 | 3.51671825            | 3.42554660                      | 3.62113722                        |
| 2   | 3.52347104 | 3.50921376            | 3.42405504                      | 3.63043113                        |
| 3   | 3.50835028 | 3.51863772            | 3.48882321                      | 3.62117951                        |
| 4   | 3.50525342 | 3.51858662            | 3.43421950                      | 3.60670734                        |
| 5   | 3.50967737 | 3.51790606            | 3.42857820                      | 3.60528875                        |
| avg | 3.51051611 | 3.51621248            | 3.44024451                      | 3.61694879                        |

Benchmarking was done on Fedora 37 on a Ryzen 3600, running at a fixed 3.6GHz, with each process restricted to a single CPU core and with ASLR disabled, to try and reduce variance.

We can see from these results that `mvh-narrow-newobj-api-remove-ec` is consitently faster than all the other branches. From this I conclude that the compiler optimisations for thread locals are good enough now to offset the performance hit caused by repeatedly accessing the thread local `ec` using `GET_EC`.

So I would like to merge the patches that remove the `ec` from the internal `NEWOBJ_OF` into this branch and merge it.

**Comment by eightbitraptor** [comment]
One thing I don't have a good answer for at this time is why `mvh-narrow-newobj-api-macro-branch` is slower than `mvh-narrow-newobj-api`, given that it's the same functionality but missing a branch.

We can even see this by looking at the generated code for the `rb_ec_str_ressurrect` function with `objdump`:

```
objdump -M intel --disassemble=rb_ec_str_resurrect --no-addresses miniruby | wc -l
```

* on the `mvh-narrow-newobj-api` the generated asm is 193 lines
* on the `mvh-narrow-newobj-api-macro-branch` branch the generated asm is 181 lines

In addition, in `mvh-narrow-newobj-api` we can clearly see the calls to `rb_current_ec` where it gets the thread local

```
❯ objdump -M intel --disassemble=rb_ec_str_resurrect --no-addresses miniruby | grep ruby_current_ec
        48 8b 05 61 5f 2b 00    mov    rax,QWORD PTR [rip+0x2b5f61]        # <ruby_current_ec@@Base+0x50dca0>
        48 8b 05 29 5f 2b 00    mov    rax,QWORD PTR [rip+0x2b5f29]        # <ruby_current_ec@@Base+0x50dca0>
```

So given that it generates more code, and that extra code is fetching the `ec`, I can't explain why it's faster.

I don't think this is anything to worry about, given that removing the `ec` is faster than both of these branches. But it's a curiosity I can't explain.

**Comment by tenderlove** [comment]
> We can see from these results that `mvh-narrow-newobj-api-remove-ec` is consitently faster than all the other branches. From this I conclude that the compiler optimisations for thread locals are good enough now to offset the performance hit caused by repeatedly accessing the thread local `ec` using `GET_EC`.

We should probably try this benchmark with more than 1 thread live in the system.  I'd guess we'll see stalls when there are multiple threads?

**Comment by eightbitraptor** [comment]
@tenderlove It looks like you're right. I think restricting the benchmark too far was skewing the results.

I changed the benchmark script to this:

```
eval("def string_literal; foo = [#{ 10_000.times.map { "'test'" }.join(",") }]; nil; end")

before = Process.clock_gettime(Process::CLOCK_MONOTONIC)

5.times.map do
  Thread.new do
    10000.times { string_literal }
  end
end.map(&:join)

p Process.clock_gettime(Process::CLOCK_MONOTONIC) - before
```

And ran all the benchmarks again, outside of the `cgroup` so they can be moved between cores. And the numbers I got were as follows (averages of 20 runs):

| feature             | time        |
| ------------------- | ----------- |
| refactor            | 16.18893562 |
| refactor+macro      | 15.64237013 |
| refactor+ec-removal | 17.37436765 |

So, as expected, this time removing the explicit `ec` and relying on `GET_EC` is the slowest, and there does seem to be a decent speedup by removing the branch from the code, and relying on the pre-processor, which makes sense as there's fewer machine instructions generated on this branch.

If you're happy with this, then I'll tidy up the commits and merge the refactor with the macro.

**Comment by tenderlove** [comment]
> If you're happy with this, then I'll tidy up the commits and merge the refactor with the macro.

Sounds good to me!

**Comment by ko1** [comment]
> We should probably try this benchmark with more than 1 thread live in the system. I'd guess we'll see stalls when there are multiple threads?

Sorry why does it introduce stalls?
